src/main/java/microsoft/exchange/webservices/data/ExchangeServiceBase.java

```diff
@@ -471,28 +479,35 @@ protected Date convertUniversalDateTimeStringToDate(String dateString) {
     if (dateString == null || dateString.isEmpty()) {
       return null;
     } else {
-      if (dateString.endsWith("Z")) {
+    TimeZone UTCTimeZone = TimeZone.getTimeZone("UTC");
```

Comment:
This is really hard to understand. If I understand it right the only difference in the parsing is the changed pattern when a `ParseException` occurs. Maybe we should have this refactored. I would suggest extracting the parsing process into another method. Smth like:
```java
private String doParsing(final boolean useUtcParser, final Date dateToBeParsed, final String pattern) 
   throws ParseException {
  if(useUtcParser) {
      	utcFormatter.setTimeZone(TimeZone.getTimeZone("UTC"));
  }
  return new SimpleDateFormat(utcPattern).parse(dateToBeParsed);
}
```
I also would suggest to let this happen in a loop rather then multiple try/catch. This loop may run until all patterns have been used or the parsing ends correctly.